kyuubi-util-scala/src/main/scala/org/apache/kyuubi/util/reflect/ReflectUtils.scala

```diff
@@ -37,32 +38,76 @@ object ReflectUtils {
       DynClasses.builder().loader(cl).impl(className).buildChecked()
     }.isSuccess
 
-  def getFieldVal[T](target: Any, fieldName: String): T =
+  private def getFieldInternal[T](target: Any, fieldName: String, isDeclaredField: Boolean): T = {
+    val targetClass = target.getClass
     Try {
-      DynFields.builder().hiddenImpl(target.getClass, fieldName).build[T]().get(target)
+      Some(DynFields.builder())
+        .map(if (isDeclaredField) {
+          _.hiddenImpl(targetClass, fieldName)
+        } else {
+          _.impl(targetClass, fieldName)
+        }).get.buildChecked[T](target).get()
     } match {
       case Success(value) => value
-      case Failure(e) =>
-        val candidates = target.getClass.getDeclaredFields.map(_.getName).mkString("[", ",", "]")
-        throw new RuntimeException(s"$fieldName not in ${target.getClass} $candidates", e)
+      case Failure(t) =>
+        val candidates = if (isDeclaredField) {
+          targetClass.getDeclaredFields
+        } else {
+          targetClass.getFields
+        }.map(_.getName).mkString("[", ",", "]")
+        throw new RuntimeException(s"$fieldName not in $targetClass $candidates", t)
     }
+  }
 
-  def getFieldValOpt[T](target: Any, name: String): Option[T] =
-    Try(getFieldVal[T](target, name)).toOption
+  def getField[T](target: Any, fieldName: String): T =
+    getFieldInternal[T](target, fieldName, isDeclaredField = false)
 
-  def invoke(target: AnyRef, methodName: String, args: (Class[_], AnyRef)*): AnyRef =
-    try {
+  def getFieldOpt[T](target: Any, name: String): Option[T] =
+    Try(getField[T](target, name)).toOption
+
+  def getDeclaredField[T](target: Any, fieldName: String): T =
+    getFieldInternal[T](target, fieldName, isDeclaredField = true)
+
+  def getDeclaredFieldOpt[T](target: Any, name: String): Option[T] =
+    Try(getDeclaredField[T](target, name)).toOption
+
+  private def invokeInternal[T](
+      target: AnyRef,
+      methodName: String,
+      isDeclaredMethod: Boolean,
+      args: (Class[_], AnyRef)*): T = {
+    val targetClass = target.getClass
+    Try {
       val (types, values) = args.unzip
-      DynMethods.builder(methodName).hiddenImpl(target.getClass, types: _*).build()
-        .invoke(target, values: _*)
-    } catch {
-      case e: NoSuchMethodException =>
-        val candidates = target.getClass.getMethods.map(_.getName).mkString("[", ",", "]")
-        throw new RuntimeException(s"$methodName not in ${target.getClass} $candidates", e)
+      Some(DynMethods.builder(methodName))
+        .map(if (isDeclaredMethod) {
+          _.hiddenImpl(targetClass, types: _*)
+        } else {
+          _.impl(targetClass, types: _*)
```

Comment:
what if we always call `hiddenImpl`, then we don't need to expose `isDeclared*` concept to caller

Reply:
Let me have a try on this.